File: app/src/main/java/ru/lionzxy/bmstuwifi/authentificator/BMSTUStudentAuth.java
package ru.lionzxy.bmstuwifi.authentificator;

import ru.lionzxy.bmstuwifi.tasks.AuthTask;
import ru.lionzxy.bmstuwifi.interfaces.ITask;
import ru.lionzxy.bmstuwifi.utils.logs.Logger;

/**
 * Created by lionzxy on 05.11.16.
 */
public class BMSTUStudentAuth extends IAuth {
    private ITask authTask = null;
    public static final String STUDENT_AUTH_SITE = "https://lbpfs.bmstu.ru:8003/index.php?zone=bmstu_lb";

    public BMSTUStudentAuth(Logger logger) {
        super(logger, "lb", "bmstu_lb");
    }

    @Override
    public void stop() {
        if (authTask != null)
            authTask.interrupt();
    }

    @Override
    public ITask registerInNetwork() {
        return new AuthTask(this);
    }

    @Override
    public String getAuthSite() {
        return STUDENT_AUTH_SITE;
    }

    @Override
    public String getLogoutSite() {
        return STUDENT_AUTH_SITE;
    }
}

File: app/src/main/java/ru/lionzxy/bmstuwifi/authentificator/provides/BMSTUStudentAuth.kt
package ru.lionzxy.bmstuwifi.authentificator.provides

import android.content.Context
import okhttp3.FormBody
import okhttp3.Request
import org.jsoup.Jsoup
import ru.lionzxy.bmstuwifi.R
import ru.lionzxy.bmstuwifi.authentificator.NamedTask
import ru.lionzxy.bmstuwifi.authentificator.Provider
import java.util.*

class BMSTUStudentAuth(context: Context) : Provider(context) {
    companion object {
        const val STUDENT_AUTH_SITE = "https://lbpfs.bmstu.ru:8003/index.php?zone=bmstu_lb"
    }

    init {
        add(object : NamedTask(context.getString(R.string.notification_connection_check)) {
            override fun run(vars: HashMap<String, Any>): Boolean {
                if (isConnected) {
                    vars["result"] = RESULT.ALREADY_CONNECTED
                    return false
                }
                return true
            }
        })

        add(object : NamedTask(context.getString(R.string.notification_register_check)) {
            override fun run(vars: HashMap<String, Any>): Boolean {
                setLogin("knyu16u287")
                setPassword("5ka3kn7w")

                if (getLogin("") == "" || getPassword("") == "") {
                    vars["result"] = RESULT.NOT_REGISTERED
                    // TODO
                    return true
                }
                return true
            }
        })

        add(object : NamedTask(context.getString(R.string.notification_register)) {
            override fun run(vars: HashMap<String, Any>): Boolean {
                val formBody = FormBody.Builder()
                        .add("redirurl", "/")
                        .add("auth_user", getLogin(""))
                        .add("auth_pass", getPassword(""))
                        .add("accept", "Continue")
                        .build()
                val request = Request.Builder()
                        .url(STUDENT_AUTH_SITE)
                        .post(formBody)
                        .build()

                val response = client.newCall(request).execute()

                val html = response.body()!!.string()
                val doc = Jsoup.parse(html)
                val logout_id = doc.select("input[name=logout_id]").firstOrNull()?.attr("value")

                if (response.isSuccessful && !logout_id.isNullOrEmpty()) {
                    setLogoutId(logout_id!!)
                    vars["result"] = RESULT.CONNECTED
                } else {
                    vars["result"] = RESULT.NOT_REGISTERED
                }
                return response.isSuccessful
            }
        })
    }
}